### PR TITLE
Add Events API/Support for tt_metal for host->CQ and CQ<->CQ synchronization #5529

### PR DESCRIPTION
**Pull request by kmabeeTT**
Changes are functional and cleaned up, got two things to follow up on

- [x] Still WIP a FIXME in code for known bug for 2 hw cq, and need to test it (will update this PR code tomorrow)
-  multi-device won't work in current implementation, no API exists for chip-to-chip reads from kernel, need to revisit soon (after this merge).

Tag @tooniz for initial review in the meantime if you have time.

### CODE REVIEW

File: tt_metal/impl/dispatch/command_queue.cpp
Comment (tooniz) [inline tt_metal/impl/dispatch/command_queue.cpp:1248]: Is this restriction necessary? Why can't we support recording events on CQ1?

File: tt_metal/impl/dispatch/command_queue.cpp
Comment (tooniz) [inline tt_metal/impl/dispatch/command_queue.cpp:1267]: Same, is this restriction necessary? Why can't we support recording events on CQ1?

**Review by tooniz — APPROVED** [review]
Feedback regarding CQ1 not blocking this merge, but it'll need to be addressed before doing multi-CQ testing with events

File: tt_metal/impl/dispatch/command_queue.cpp
Comment (kmabeeTT) [reply tt_metal/impl/dispatch/command_queue.cpp:1248]: Not necessary, but keeping until I can test with multi-CQ (WIP today, hope to include with this PR).

File: tt_metal/impl/dispatch/command_queue.cpp
Comment (kmabeeTT) [reply tt_metal/impl/dispatch/command_queue.cpp:1267]: Not necessary, but keeping until I can test with multi-CQ (WIP today, hope to include with this PR).

**Comment by kmabeeTT** [comment]
Hi @TT-billteng , @pgkeller , @abhullar-tt this is all tidied up, can I get your approval on these changes for #5529 please?  This is initial implementation of 3 Events API's to be used for inserting and syncing on Events from Host or Device (CQ).

- implementation, tested with synchronous hw CQ's (num_hw_cq=1, 2) for single device
- new gtests (still WIP on another 1 or 2 to include before this gets merged)
- remove redundant tt_eager APIs with same scope/name as tt_metal after chat with Akhmed.

Restriction (going to add this comment in parent ticket):

- Won't work with multi-device today, added an assert to catch if user uses `EnqueueQueueRecordEvent(cq0..)` and `EnqueueQueueWaitForEvent(cq1...)` where cq0 and cq1 are on two different devices.  Needs to be discussed next, and revisited after multi-device FD makes bit more progress. May need to change to have "last_completed_event" stored in hugepage per CQ (but strongly prefer not to, due to perf impact).

Some thoughts/decisions along the way in case it helps with review:

- matches interface discussed in parent ticket #5529
- I purposely kept this implementation simple, light. Thought about having a program/kernel to do the "on device wait" but realized it's just a couple lines of code to have the while-loop in the dispatcher kernel.
- To facilitate "Event Sync" I just pass 4 args to dispatcher kernel in 8 bytes (event_id, noc_x, noc_y, sync_ena) to tell dispatcher that this is a sync cmd and it should poll some noc_xy core at known address (reserved for "last completed event ID") until desired event_id (or greater) is seen.
- Reserved an extra slot in common runtime adress map in the gap that is used for dispatch core to store "last completed event_id" for that sync to work.
- Similarly, on host, added an extra state variable per CQ to store "last completed event id" for the host side sync.

File: tests/tt_metal/tt_metal/unit_tests_fast_dispatch_single_chip_multi_queue/command_queue/test_EnqueueQueueWaitForEvent.cpp
Comment (pgkeller) [inline tests/tt_metal/tt_metal/unit_tests_fast_dispatch_single_chip_multi_queue/command_queue/test_EnqueueQueueWaitForEvent.cpp:50]: 3, 4 made me pause, both are <10, maybe use constexpr for them and assert they are less than num_events?

File: tests/tt_metal/tt_metal/unit_tests_fast_dispatch_single_chip_multi_queue/command_queue/test_EnqueueQueueWaitForEvent.cpp
Comment (pgkeller) [inline tests/tt_metal/tt_metal/unit_tests_fast_dispatch_single_chip_multi_queue/command_queue/test_EnqueueQueueWaitForEvent.cpp]: tests above appear to cover well that events happen in order alone and across queues, but not clear we have a test that ensures that events wait until, eg, a kernel completes.  wonder if you could run something that runs long and returns some data and then afterwards do a host write to trash that location of L1.  if the event completes early, then the data returened be corrupt.  or maybe this is covered somehow?

File: tests/tt_metal/tt_metal/unit_tests_fast_dispatch_single_chip_multi_queue/command_queue/test_EnqueueQueueWaitForEvent.cpp
Comment (pgkeller) [inline tests/tt_metal/tt_metal/unit_tests_fast_dispatch_single_chip_multi_queue/command_queue/test_EnqueueQueueWaitForEvent.cpp:167]: ok, this test covers ordering, cool

**Review by pgkeller — APPROVED** [review]
recommend review by Almeet and/or Andrew

File: tt_metal/host_api.hpp
Comment (abhullar-tt) [inline tt_metal/host_api.hpp:429]: For docs to pick up these APIs you will need to add them under `docs/source/tt_metal/apis/host_apis/command_queue/`

File: tt_metal/impl/dispatch/command_queue.cpp
Comment (abhullar-tt) [inline tt_metal/impl/dispatch/command_queue.cpp:982]: I think this should be checking that issue_queue_write_ptr + command_size >= issue queue limit

File: tt_metal/impl/dispatch/command_queue.cpp
Comment (abhullar-tt) [inline tt_metal/impl/dispatch/command_queue.cpp:995]: same as above, should be checking against issue queue limit

File: tt_metal/impl/dispatch/kernels/command_queue_consumer.hpp
Comment (abhullar-tt) [inline tt_metal/impl/dispatch/kernels/command_queue_consumer.hpp:310]: the `_SCRATCH` suffix is a little misleading because it isn't really being used as scratch space, it seems to be a sync location so maybe this can be renamed to `CQ_COMPLETION_SYNC`?

**Review by abhullar-tt — CHANGES_REQUESTED** [review]

File: tt_metal/impl/dispatch/kernels/command_queue_consumer.hpp
Comment (kmabeeTT) [reply tt_metal/impl/dispatch/kernels/command_queue_consumer.hpp:310]: We talked offline, I'm treating it as a scratch region that could be used as destination for any type of 16B noc read to L1 by this dispatcher core, but just using it here for the first time to read back the last-completed-event-id.

File: tt_metal/host_api.hpp
Comment (kmabeeTT) [reply tt_metal/host_api.hpp:429]: Whoops yes thanks good call - made updates and tested them out locally with web server.

File: tt_metal/impl/dispatch/command_queue.cpp
Comment (kmabeeTT) [reply tt_metal/impl/dispatch/command_queue.cpp:982]: Whoops yes I copied buggy version from enqueue-program path, thx for flagging, fixed, about to push.

File: tests/tt_metal/tt_metal/unit_tests_fast_dispatch_single_chip_multi_queue/command_queue/test_EnqueueQueueWaitForEvent.cpp
Comment (kmabeeTT) [reply tests/tt_metal/tt_metal/unit_tests_fast_dispatch_single_chip_multi_queue/command_queue/test_EnqueueQueueWaitForEvent.cpp]: Yep, I have other tests below and newly added after this and WIP that will further test multi-cq ordering, this one was purposely kept ~simple.

File: tests/tt_metal/tt_metal/unit_tests_fast_dispatch_single_chip_multi_queue/command_queue/test_EnqueueQueueWaitForEvent.cpp
Comment (kmabeeTT) [reply tests/tt_metal/tt_metal/unit_tests_fast_dispatch_single_chip_multi_queue/command_queue/test_EnqueueQueueWaitForEvent.cpp:50]: Yeah, that's fair. Not worth the trouble to come up with some funky expression, I just changed it to loop over all events per CQ again, and sync on them in order per CQ (instead of bouncing between CQs), pushing in few mins after rebase/build tested locally.

**Review by abhullar-tt — APPROVED** [review]

**Comment by kmabeeTT** [comment]
Passing CI before merge at TOTT:

https://github.com/tenstorrent-metal/tt-metal/actions/runs/8089432962

**Pull request merged**
